[Markus Lim Yong Sheng] iP

markuslim24 opened the pull request:
# Brobot :punch:
> Your very own brother from another computer to help you manage your tasks!

## Features
* User-Friendly: Brobot communicates purely through the CLI with very simple commands. Brobot will also always be your bro!
* Fast: Brobot knows how important time is to you, so he does things __FAST__
* Local save: Brobot will still remember your tasks even after you quit the application. _What a bro move_ :punch:

## Installation
1. Download the latest release from [here](https://github.com/markuslim24/ip/releases).
2. Double click the jar file to run it.(Alternatively, run java -jar [application].jar from CLI)
3. Start typing away!

## Implementation list:
- [x] Basic task creation
    * Todo
    * Events
    * Deadline
- [x] Mark done & delete task function
- [x] Task list local save
- [ ] More commands
- [ ] Better UI(A GUI)

***

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Brobot {
    public static void main(String[] args) {

            Storage storage = new Storage("./data/list1.txt");
            TaskList list = storage.readList();
            BroParser parser = new BroParser(list, storage);

            UI.printGreeting();
            String input = UI.getUserInput();

            while (!input.equals("bye")) {
                try {
                    parser.parse(input);
                } catch (BroException e) {
                    UI.printError(e);
                } finally {
                    input = UI.getUserInput();
                }
            }
            UI.printBye();
    }
}
```

Commit:
Add Gradle support

markuslim24 committed:
level1-Greet,Echo,Exit

markuslim24 committed:
level2-Add,List

markuslim24 committed:
implemented stop message when max list size is hit

markuslim24 committed:
Modified Task to be its own class and modified list output to show if task is completed

markuslim24 committed:
Implemented done keyword functionality

markuslim24 committed:
added deadline and event classes

markuslim24 committed:
fixed some bugs in Level-3(single word inputs not registered)

markuslim24 committed:
added Todo class and implemented todo, deadline and event keyword recognition

markuslim24 committed:
abstracted task list into new TaskList class

markuslim24 committed:
modified message for unrecognized input

markuslim24 committed:
organised classes into packages and created automated testcase

markuslim24 committed:
added exception handling for invalid commands, invalids command parameter and no such tasks error(when marking tasks as done)

markuslim24 committed:
modified test case to handle new error messages

markuslim24 committed:
updated TaskList class to use ArrayList to store tasks instead of a simple limited Task Array

markuslim24 committed:
implemented delete functionality

markuslim24 committed:
updated test case to also test delete command

markuslim24 committed:
create a UI class to handle terminal input/output

markuslim24 committed:
create a parser class to process user input and execute the intended action

markuslim24 committed:
remove unnecessary ui package and move UI class directly to the root directory of Duke package

markuslim24 committed:
implement printError function on UI so UI handles printing of exception messages to terminal

markuslim24 committed:
create toStorageString function in tasklist to output a string version of itself for storage

markuslim24 committed:
write tasklist storage string to savefile after any change to list

markuslim24 committed:
read tasklist from storage

markuslim24 committed:
Merge branch 'branch-Level-7'

markuslim24 committed:
refactor package to follow coding standard

markuslim24 committed:
refactor code to follow coding standard

markuslim24 committed:
implement date parsing in parser and format output of task dates

markuslim24 committed:
Merge branch 'branch-Level-8'

markuslim24 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2122S1/ip

markuslim24 committed:
add JUnit tests

markuslim24 committed:
create jar file of app with gradle shadowJar plugin

markuslim24 committed:
add javadocs for exception package

markuslim24 committed:
add javadocs for main duke class

markuslim24 committed:
add javadocs for UI class

markuslim24 committed:
add javadocs for Storage class

markuslim24 committed:
add javadocs for Task package

markuslim24 committed:
add javadocs for DukeParser class

markuslim24 committed:
Merge branch 'branch-A-JavaDoc'

markuslim24 committed:
Implement find command.

Find for tasks matching a specified search word and print it out as a
list.

markuslim24 committed:
Merge branch 'branch-Level-9'

markuslim24 committed:
removed jar file from root directory

markuslim24 committed:
Rename entire program to brobot

The program name and theme is Duke

Duke is the original brownfield project name which is very unoriginal

lets change the program name and messages to fit a new name: Brobot.

markuslim24 committed:
Edit readme file with description of brobot

markuslim24 committed:
add checkstyle-IDEA plugin and configure plugin with checkstyle.xml file

markuslim24 committed:
reformat code to follow checkstyle using checkstyle-IDEA plugin

markuslim24 committed:
Merge branch 'branch-A-CheckStyle'

markuslim24 committed:
update build.gradle file to include javafx dependancies

markuslim24 committed:
Setup launcher and initialize application for javaFX

markuslim24 committed:
add GUI mockup from tutorial

markuslim24 committed:
implement broParser into GUI to handle user input and respond correctly

markuslim24 committed:
clip display pictures to be a circle

markuslim24 committed:
add padding to dialogbox

markuslim24 committed:
Merge branch 'branch-Level-10'

markuslim24 committed:
polish up javadocs comments and modify build.gradle file to use Launcher class as entry point

imposterble commented on src/test/java/brobot/parser/BroParserTest.java:
```diff
@@ -0,0 +1,37 @@
+package brobot.parser;
+
+import brobot.task.TaskList;
+import brobot.Storage;
+import brobot.exception.BroException;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
```
Should the order of the import statements be like this?

```java
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;

import brobot.Storage;
import brobot.exception.BroException;
import brobot.task.TaskList;````

imposterble commented on src/test/java/brobot/parser/BroParserTest.java:
```diff
@@ -0,0 +1,37 @@
+            p.parse("hello");
+            fail();
+        } catch (BroException e) {
+            assertEquals("Oops, unfortunately i am not yet smart enough to understand what you are saying", e.getMessage());
```
Perhaps this line can be split into two lines?

imposterble commented on src/test/java/brobot/task/TaskListTest.java:
```diff
@@ -0,0 +1,22 @@
+package brobot.task;
+
+import brobot.exception.BroException;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
Should the order of the import statements be like this?

```java
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;

import brobot.exception.BroException;```

imposterble commented on src/test/java/brobot/UITest.java:
```diff
@@ -0,0 +1,9 @@
+package brobot;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
```
Perhaps these unused lines can be removed?

imposterble commented on src/test/java/brobot/UITest.java:
```diff
@@ -0,0 +1,9 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class UITest {
```
Should the class name be UiTest?

imposterble commented on src/main/java/brobot/Storage.java:
```diff
@@ -0,0 +1,102 @@
+            while (s.hasNext()) {
+                String[] arr = s.nextLine().split(" \\| ", 4);
+                String taskType = arr[0];
+                switch (taskType) {
```
I like your usage of the switch case statement here.

kdeasymoneysniper replied:
Yeah! a switch statement works great here!

imposterble commented on src/main/java/brobot/UI.java:
```diff
@@ -0,0 +1,131 @@
+package brobot;
```
I like how you packaged the classes with similar functionalities into sub-packages!

imposterble commented on src/main/java/brobot/UI.java:
```diff
@@ -0,0 +1,131 @@
+ * Represents the User interface of the Duke Program. Interfaces through the terminal.
+ */
+public class UI {
+    protected static String greetingText = "Hey Bro! I'm Brobot\nWhat can I do for you bro?\n";
```
I like the greeting text of your bot 😃

markuslim24 replied:
Thanks bro 👊

imposterble reviewed (commented):
LGTM! Clean code 👍🏻 Well done and all the best with the remaining iP tasks 😃

kdeasymoneysniper commented on src/main/java/brobot/UI.java:
```diff
@@ -0,0 +1,131 @@
+     * @param list The list from which the task was deleted.
+     */
+    public static String printTaskDeleted(Task task, TaskList list) {
+        System.out.println(
+                formatWithSpace("Ok Bro, I removed the following task:\n "
+                    + task
+                    + listTaskAmount(list.getTaskAmount()))
+        );
+        return formatWithSpace("Ok Bro, I removed the following task:\n "
+                + task
+                + listTaskAmount(list.getTaskAmount()));
```
Nice separation of lines for readability!

kdeasymoneysniper commented on src/main/java/brobot/exception/NoSuchTaskException.java:
```diff
@@ -0,0 +1,13 @@
+     * Constructor for the exception.
+     */
+    public NoSuchTaskException() {
+        super("Ey bro, no such task lah!");
```
Your exceptions are really clear and it is split up properly. And the added humour is greatly appreciated!

kdeasymoneysniper reviewed (commented):
Code was enjoyable to read, and it is well organised! :-)

markuslim24 committed:
Refactor UI class functions to be more basic and reusable

Most functions in the UI class handle both the string formatting of
the text and printing it.

The GUI does not need to print the text to the terminal, it just needs
the text.

I will separate out the print to Terminal functionality into a separate
function so the main functions will only return the formatted text.

markuslim24 committed:
Merge branch 'branch-A-CodingStandard'

markuslim24 committed:
Tasklist, Storage and Parser class: Add assertions

The above-mentioned classes make unchecked assumptions such as the
taskNUmber being more than 0 and the filePath being writable in storage

Should these criterias not be checked and indeed fail, there could be serious
bugs occuring during runtime.

I will add assertions to check for the assumptions made in Tasklist,
Storage and Parser class to ensure that the program stops if these
assumptions are not met.

markuslim24 committed:
Merge pull request #1 from markuslim24/branch-A-Assertions

Tasklist, Storage and Parser class: Add assertions

markuslim24 committed:
JUnit Tests: remove unsued test cases and refactor test classes to
follow checkStyle

There are unused JUnit test classes and the ones that are necessary do
not follow the checkStyle

Resolving these issues will clean up the code and make it more
reader-friendly

I will remove the unused classes and refactor the test cases to follow
checkStyle

markuslim24 committed:
Separate main Brobot class into BrobotGui and Brobot Cli

markuslim24 committed:
Improve code quality in UI class

Some functions and variables in the UI class were not reader-friendly

Refactoring the code to be more reader-friendly would improve code
quality.

I will refactor the code to make it more readable.

markuslim24 committed:
improve code quality in Storage class

markuslim24 committed:
improve code quality in classes under task package

markuslim24 committed:
Merge pull request #2 from markuslim24/branch-A-CodeQuality

Branch a code quality

markuslim24 committed:
Add gradle.yml file for continuous integration

markuslim24 committed:
Update JUnit tests to to handle new exception messages for brobot

The current JUnit tests still test for the exception messages when it
was a Duke program

This change to test the new messages will fix the tests

I will edit the test to ensure it is testing for the correct message

markuslim24 committed:
update TaskList test to handle exception messages from brobot

markuslim24 committed:
Storage command: type 'storage' to change storage directory

The current storage directory is fixed to ./data and cannot be changed
by the user.

Allowing the user to change the directory will allow more flexibility
for the user to choose where he/she wants to store their data

I will implement the storage command to open a GUI for which the user
can select their new save directory. The directory will be remembered
even after the application is closed.

markuslim24 committed:
Clean up code and improve code quality for storage command

The current implementation for storage command has missing javadoc
comments and messy code

Cleaning up the code will improve code quality and make it more readable

I will package the storage classes together, clean up unnecessary code
and input the missing javadocs

markuslim24 committed:
removed some temp directories for push to repo

markuslim24 committed:
rearrange import statements to pass checkstyle

markuslim24 committed:
rearrange import statement again to pass checkstyle

markuslim24 committed:
Merge pull request #3 from markuslim24/branch-C-FlexibleDataSource

Branch c flexible data source

markuslim24 committed:
separate formatting function from messages for cleaner GUI outputs

markuslim24 committed:
wrap reply in text bubble and improve padding

markuslim24 committed:
minor padding tweaks to gui

markuslim24 committed:
add greeting to Gui upon first load

markuslim24 committed:
Set theme jekyll-theme-cayman

markuslim24 committed:
add Ui.png to docs

markuslim24 committed:
remove unnecessary code in custom DialogBox class

markuslim24 committed:
Merge branch 'master' of https://github.com/markuslim24/ip

markuslim24 committed:
update readme

markuslim24 committed:
update pages config.yml

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
update readme

markuslim24 committed:
include photo credits

markuslim24 committed:
reduce photo credit line to less than 120 words
